### PR TITLE
[baremetal] Merge Mbed TLS 2.16 (2019-09-09)

### PR DESCRIPTION
**Pull request by mpg**
## Description

This is our periodic merge of the 2.16 branch of Mbed TLS into the baremetal branch.

This edition comes with a fix to one source of spurious failures in the CI, that will be quite welcome. It also came with a few merge conflicts:

* the fix for HMAC-DRBG entropy collection turned out to be sligthly different on both sides - I kept the version with one more comment
* 2.16 has a fix for an compiler warning in X.509 (#2813) that didn't apply cleanly on baremetal as the name of one variable was changed (`parent` -> `parent_crt`) during on-demand parsing work. I applied the fix as in 2.16 but keeping the new name.
* 2.16 fixed a bug in writing EC keys (https://github.com/ARMmbed/mbedtls-restricted/pull/503), while baremetal reworked writing EC keys when adding the TinyCrypt option. This resulted in two conflicts: a trivial one in the test suite (resolved in the merge commit), and a less trivial one in the library (resolved in a follow-up commit, see that commit's message for details).
* [new with latest baremetal] 2.16 added an EC J-PAKE test that uses the MD module in a way that needed changing to work with Hanno's recent work on hardcoding a single hash in the MD layer - done in a separate commit.

## Status
**READY**

### CODE REVIEW

File: ChangeLog [hunk 1]
```diff
@@ -47,6 +72,9 @@ API Changes
      fields of `mbedtls_x509_crt` are removed, but it keeps the
      X.509 function API. See the API changes section as well as
      the documentation in `config.h` for more information.
+   * The new function mbedtls_ecdsa_sign_det_ext() is similar to
```
Comment (hanno-arm) [inline ChangeLog]: This was in the API changes column in 2.16 - why did it move?

Comment (hanno-arm) [reply ChangeLog]: Oh, wait, I'm mixing something up. Nevermind!

**Comment by mpg** [comment]
@jarlamsa Thanks for your review and catching the issue in the ChangeLog. I had a look and the issue actually much bigger than that, ChangeLog entries from 2.16 and baremetal are all intermixed. I'm working on unmixing them and will push when I'm done.

**Comment by mpg** [comment]
@jarlamsa @hanno-arm I believe I fixed the ChangeLog. Please review again.

**Comment by mpg** [comment]
Travis is actually passing but just failed to report status to github: https://travis-ci.org/ARMmbed/mbedtls/builds/584023036 - removing "needs: CI".

**Comment by mpg** [comment]
@hanno-arm @jarlamsa This is a kind reminder that this PR has been awaiting your re-review for a week now. It passes CI (disregarding one spurious failure) and each of you approved a previous version, so hopefully it should be quite ready now.

**Comment by jarlamsa** [comment]
Actually the CI-job seems to be utterly failing (because of itself) and not just one spurious test failure.

**Comment by mpg** [comment]
@jarlamsa Turns out the report was from an experimental version of the CI job. Since it was using the same name as the actual job when reporting to github, the results were confusing. I did a zero-diff force-push in order to discard it and start a new, fresh run of (hopefully just) the actual PR jobs.

**Comment by jarlamsa** [comment]
Couple of spurious failures, otherwise passing

**Comment by mpg** [comment]
@hanno-arm @jarlamsa Thanks for your quick reaction to my reminder :)

Jenkins is passing, travis is [passing as well](https://travis-ci.org/ARMmbed/mbedtls/builds/586939627) even if reporting to github is currently broken, so with two approvals, a passing CI and no conflicts, this is ready for merge.

**Pull request merged**